sys/net/gnrc/transport_layer/tcp/gnrc_tcp_eventloop.c

```diff
@@ -0,0 +1,259 @@
+    if (tcp == NULL) {
+        DEBUG("gnrc_tcp_eventloop : _send() : tcp header missing.\n");
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    /* Dispatch to network layer */
+    if (!gnrc_netapi_dispatch_send(pkt->type, GNRC_NETREG_DEMUX_CTX_ALL, pkt)) {
+        DEBUG("gnrc_tcp_eventloop : _send() : network layer not found\n");
+        gnrc_pktbuf_release(pkt);
+    }
+    return 0;
+}
+
+/**
+ * @brief recv function, used to call fsm on packet reception
+ *
+ * @param[in] pkt   incomming paket to process
+ *
+ * @return zero on success
+ * @return negative value on error
+ * @return -EACCES if not able to get write access to packet
+ * @return -ERANGE if segment offset is less than 5
+ * @return -ENOMSG if paket can't be marked
+ * @return -EINVAL if checksum was invalid
+ * @return -ENOTCONN if no module is interested in this context
+ */
+static int8_t _receive(gnrc_pktsnip_t *pkt)
+{
+    /* NOTE: Receiving direction: pkt = payload, payload->next = tcp, tcp->next = nw */
+    uint16_t ctl = 0;
+    uint16_t src = 0;
+    uint16_t dst = 0;
+    uint8_t hdr_size = 0;
+    uint8_t syn = 0;
+
+    gnrc_pktsnip_t *tcp = NULL;
+    gnrc_pktsnip_t *ip6 = NULL;
+    gnrc_pktsnip_t *rst = NULL;
+    conn_tcp_t *conn = NULL;
+    tcp_hdr_t *hdr;
+
+    /* mark TCP header */
+    tcp = gnrc_pktbuf_start_write(pkt);
+    if (tcp == NULL) {
+        DEBUG("gnrc_tcp_eventloop : _receive() : can't write to packet\n");
+        gnrc_pktbuf_release(pkt);
+        return -EACCES;
+    }
+    pkt = tcp;
+
+    /* Get IP Header */
+    LL_SEARCH_SCALAR(pkt, ip6, type, GNRC_NETTYPE_IPV6);
+    assert(ip6 != NULL);
+
+    /* Get TCP Header */
+    LL_SEARCH_SCALAR(pkt, tcp, type, GNRC_NETTYPE_TCP);
+    assert(tcp != NULL);
+
+    /* Extract control bits */
+    ctl = byteorder_ntohs(((tcp_hdr_t *) tcp->data)->off_ctl);
+
+    /* Validate Offset */
+    if (GET_OFFSET(ctl) < OPTION_OFFSET_BASE) {
+        DEBUG("gnrc_tcp_eventloop : _receive() : unexpected Offset Value\n");
+        gnrc_pktbuf_release(pkt);
+        return -ERANGE;
+    }
+
+    /* Calculate tcp header size */
+    hdr_size = GET_OFFSET(ctl) * 4;
+
+    /* Mark TCP-Header, if it contains any payload */
+    if ((pkt->type == GNRC_NETTYPE_TCP) && (pkt->size != hdr_size)) {
+        tcp = gnrc_pktbuf_mark(pkt, hdr_size, GNRC_NETTYPE_TCP);
+        if (tcp == NULL) {
+            DEBUG("gnrc_tcp_eventloop : _receive() : Header marking failed\n");
+            gnrc_pktbuf_release(pkt);
+            return -ENOMSG;
+        }
+        pkt->type = GNRC_NETTYPE_UNDEF;
+    }
+
+    hdr = (tcp_hdr_t *)tcp->data;
+
+    /* Validate Checksum */
+    if (byteorder_ntohs(hdr->checksum) != _pkt_calc_csum(tcp, ip6, pkt)) {
+        DEBUG("gnrc_tcp_eventloop : _received() : Invalid checksum\n");
+        gnrc_pktbuf_release(pkt);
+        return -EINVAL;
+    }
+
+    /* De-multiplex connection */
+    syn = ((ctl & MSK_SYN_ACK) == MSK_SYN) ? true : false;
```

Comment:
No Problem. The "(ctl & MSK_SYN_ACK)" extracts the Bits of SYN and ACK. This is to check is SYN and only SYN is set, so both must be extracted